File: Blind150/6.LinkedList/GO/copy_list_with_random.go
// https://leetcode.com/problems/copy-list-with-random-pointer/description/
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Next *Node
 *     Random *Node
 * }
 */

func copyRandomList(head *Node) *Node {

	if head == nil {
		return nil
	}

	OldToCopy := make(map[*Node]*Node)

	cur := head

	for cur != nil {

		new := Node{Val: cur.Val}
		OldToCopy[cur] = &new
		cur = cur.Next

	}

	cur = head

	for cur != nil {

		copied := OldToCopy[cur]
		copied.Next = OldToCopy[cur.Next]
		copied.Random = OldToCopy[cur.Random]
		cur = cur.Next

	}

	return OldToCopy[head]

}

File: Blind150/6.LinkedList/Python/copy_list_with_random.py
# https://leetcode.com/problems/copy-list-with-random-pointer/description/

from typing import Optional
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':

        if not head:
            return None

        oldToCopy = {}

        cur = head
        while cur:

            copy = Node(cur.val)
            oldToCopy[cur] = copy
            cur = cur.next

        cur = head

        while cur:
            copied = oldToCopy.get(cur)

            copied.next = oldToCopy.get(cur.next)
            copied.random = oldToCopy.get(cur.random)
            cur = cur.next

        return oldToCopy[head]
